### DISCUSSION TITLE
Passing variable from DevOps pipeline into Elsa workflow

### DISCUSSION THREAD

**Discussion by CoyWhiting** [elsa-workflows/elsa-core#614] [Q&A] [answered]
I have a use case where we want a DevOps build pipeline to trigger a workflow in Elsa. I was going to use HTTP Request to trigger the workflow and call it from the DevOps pipeline but could use another method if needed.

So how can I pass variables from a DevOps pipeline (or any other outside source) into an Elsa workflow?

**Comment by sfmskywalker** [comment, marked as answer]
That's a great use case.

The easiest way to pass variables into an Elsa Workflows is by using the `HTTP Request Received` trigger, effectively exposing the workflow as an endpoint. You can then post anything you like from your pipeline, ideally as a JSON body. That JSON will then be stored in the workflow and made available to the rest of it.

**Reply by CoyWhiting** [reply]
UPDATE: I figured it out. I had my application running as a service and I had authentication turned on in the service to require a token.

I tried using the following through a PowerShell script.

```
$vars = @{
    file='test.zip'
}
$json = $vars | ConvertTo-Json
$response = Invoke-RestMethod 'http://xxxxxxxxxxxxx.com/testrequest' -Method POST -Body $json -ContentType 'application/json'
```

But I keep getting a 401 Unauthorized error. I have the Receive HTTP Request activity set to a method of POST. Hitting a Receive HTTP Request activity of GET works fine but obviously I can't pass in any JSON.

```
Invoke-RestMethod : The remote server returned an error: (401) Unauthorized.
At line:5 char:13
+ $response = Invoke-RestMethod 'http://xxxxxxxxxxxxx.com: ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand
```
